agents/ten_packages/system/ten_ai_base/interface/ten_ai_base/llm.py

```diff
@@ -0,0 +1,119 @@
+#
+# This file is part of TEN Framework, an open source project.
+# Licensed under the Apache License, Version 2.0.
+# See the LICENSE file for more information.
+#
+from abc import ABC, abstractmethod
+import asyncio
+from enum import Enum
+from typing import Optional, TypedDict, Union
+
+from pydantic import BaseModel
+from ten import (
+    AsyncExtension,
+    TenEnv,
+    Data,
+)
+from ten.async_ten_env import AsyncTenEnv
+from ten.cmd import Cmd
+from ten.cmd_result import CmdResult, StatusCode
+from .const import CMD_PROPERTY_TOOL, CMD_TOOL_REGISTER
+from .types import TenLLMCallCompletionArgs, TenLLMDataCompletionArgs, TenLLMToolMetadata
+from .helper import AsyncQueue
+import json
+
+class TenLLMBaseExtension(AsyncExtension, ABC):
+    # Create the queue for message processing
+    queue = AsyncQueue()
+    available_tools: list[TenLLMToolMetadata] = []
+    available_tools_lock = asyncio.Lock()  # Lock to ensure thread-safe access
+    current_task = None
+
+    async def on_init(self, ten_env: TenEnv) -> None:
+        ten_env.log_debug("on_init")
+
+    async def on_start(self, ten_env: TenEnv) -> None:
+        ten_env.log_debug("on_start")
+
+        self.loop = asyncio.get_event_loop()
+        self.loop.create_task(self._process_queue(ten_env))
+
+    async def on_stop(self, ten_env: TenEnv) -> None:
+        ten_env.log_debug("on_stop")
+
+    async def on_deinit(self, ten_env: TenEnv) -> None:
+        ten_env.log_debug("on_deinit")
+
+    async def on_cmd(self, async_ten_env: AsyncTenEnv, cmd: Cmd) -> None:
+        cmd_name = cmd.get_name()
+        async_ten_env.log_debug(f"on_cmd name {cmd_name}")
+        try:
+            if cmd_name == CMD_TOOL_REGISTER:
+                tool_metadata_json = json.loads(cmd.get_property_to_json(CMD_PROPERTY_TOOL))
+                async_ten_env.log_info(f"register tool: {tool_metadata_json}")
+                tool_metadata = TenLLMToolMetadata.model_validate_json(tool_metadata_json)
+                async with self.available_tools_lock:
+                    self.available_tools.append(tool_metadata)
+                await self.on_tools_update(async_ten_env, tool_metadata)
+                async_ten_env.return_result(CmdResult.create(StatusCode.OK), cmd)
+        except Exception as err:
+            async_ten_env.log_warn(f"on_cmd failed: {err}")
+            async_ten_env.return_result(CmdResult.create(StatusCode.ERROR), cmd)
+
+
```

Comment:
what about on_data? and what about the extension template?

Reply:
no_data is not complicated. so i suggest to provide default implementation in template.
template is difinitely needed but it should be put into TEN？